File: TelerikUIExamplesInCSharp/Examples/DataForm/DataFormCustomization.cs
using System;
using TelerikUI;
using UIKit;
using Foundation;
using System.Collections.Generic;
using CoreGraphics;

namespace Examples
{
	[Register("DataFormCustomization")]
	public class DataFormCustomization : TKDataFormViewController
	{
		CustomizationDataFormDelegate dataFormDelegate;
		TKDataFormEntityDataSourceHelper dataSource;
		ReservationForm reservationForm;
		UIButton btn;

		public override void ViewDidLoad ()
		{
			base.ViewDidLoad ();

			this.View.BackgroundColor = UIColor.White;

			this.reservationForm = new ReservationForm ();
			this.dataSource = new TKDataFormEntityDataSourceHelper (this.reservationForm);
			this.dataFormDelegate = new CustomizationDataFormDelegate ();

			NSDateFormatter formatter = new NSDateFormatter ();
			formatter.DateFormat = "h:mm a";
			this.dataSource.PropertyWithName ("Time").Formatter = formatter;

			this.dataSource["Name"].Image = UIImage.FromBundle ("guest-name.png");
			this.dataSource["Phone"].Image = UIImage.FromBundle ("phone.png");
			this.dataSource["Date"].Image = UIImage.FromBundle ("calendar.png");
			this.dataSource["Time"].Image = UIImage.FromBundle ("time.png");
			this.dataSource["Guests"].Image = UIImage.FromBundle ("guest-number.png");
			this.dataSource["Table"].Image = UIImage.FromBundle ("table-number.png");

			this.dataSource["Name"].HintText = "Name";
			this.dataSource["Name"].ErrorMessage = @"Please fill in the guest name";
			this.dataSource["Time"].EditorClass = new ObjCRuntime.Class(typeof(TKDataFormTimePickerEditor));
			this.dataSource["Phone"].EditorClass = new ObjCRuntime.Class(typeof(CallEditor));
			this.dataSource ["Phone"].HintText = "Phone";
			this.dataSource["Origin"].EditorClass = new ObjCRuntime.Class(typeof(TKDataFormSegmentedEditor));

			this.dataSource["Guests"].Range = new TKRange (new NSNumber(1), new NSNumber(10));
			this.dataSource["Section"].ValuesProvider = NSArray.FromStrings (new string[] {
				"Section 1",
				"Section 2",
				"Section 3",
				"Section 4"
			});
			this.dataSource["Table"].ValuesProvider = NSArray.FromStrings(new string[] { "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15" });
			this.dataSource["Origin"].ValuesProvider = NSArray.FromStrings (new string[] {
				"phone",
				"in-person",
				"online",
				"other"
			});

			this.dataSource.AddGroup ("RESERVATION DETAILS", new string[] { "Name", "Phone", "Date", "Time", "Guests" });
			this.dataSource.AddGroup ("TABLE DETAILS", new string[] { "Section", "Table" });
			this.dataSource.AddGroup ("ORIGIN", new string[] { "Origin" });
				
			this.DataForm.BackgroundColor = UIColor.FromPatternImage (UIImage.FromBundle ("wood-pattern.png"));
			this.DataForm.Frame = new CGRect (0, 0, this.View.Bounds.Size.Width, this.View.Bounds.Size.Height - 66);
			this.DataForm.TintColor = new UIColor (0.780f, 0.2f, 0.223f, 1.0f);
			this.DataForm.Delegate = this.dataFormDelegate;
			this.DataForm.WeakDataSource = this.dataSource.NativeObject;

			btn = new UIButton (new CGRect (0, this.DataForm.Frame.Size.Height, this.View.Bounds.Size.Width, 66));
			btn.SetTitle ("Cancel Reservation", UIControlState.Normal);
			btn.SetTitleColor (new UIColor (0.780f, 0.2f, 0.223f, 1.0f), UIControlState.Normal);
			btn.AddTarget (this, new ObjCRuntime.Selector ("CancelReservation"), UIControlEvent.TouchUpInside);
			this.View.AddSubview (btn);
		}

		public override void ViewWillLayoutSubviews ()
		{
			base.ViewWillLayoutSubviews ();
			btn.Frame = new CGRect (0, this.DataForm.Frame.Size.Height, this.View.Bounds.Size.Width, 66);
		}

		[Export ("CancelReservation")]
		void CancelReservation()
		{
			TKAlert alert = new TKAlert();

			alert.Style.CornerRadius = 3;
			alert.Title = "Cancel Reservation";
			alert.Message = "Reservation Canceled!";

			alert.AddActionWithTitle ("OK", (TKAlert a, TKAlertAction action) => { return true; });
			alert.Show(true);
		}
	}
		
	class CustomizationDataFormDelegate: TKDataFormDelegate
	{
		public override bool ValidateProperty (TKDataForm dataForm, TKEntityProperty property, TKDataFormEditor editor)
		{
			if (property.Name == "Name") {
				NSString value = (NSString)property.ValueCandidate;
				if (value.Length == 0) {
					return false;
				}
			}

			return true;
		}

		public override void UpdateEditor (TKDataForm dataForm, TKDataFormEditor editor, TKEntityProperty property)
		{
			editor.Style.TextLabelOffset = new UIOffset (10, 0);
			editor.Style.SeparatorLeadingSpace = 40;
			editor.Style.AccessoryArrowStroke = new TKStroke (new UIColor (0.780f, 0.2f, 0.233f, 1.0f));
			List<string> properties = new List<string> () { "Origin", "Date", "Time", "Name", "Phone" };

			if (properties.Contains(property.Name)) {
				editor.Style.TextLabelDisplayMode = TKDataFormEditorTextLabelDisplayMode.Hidden;
				TKGridLayoutCellDefinition titleDef = editor.GridLayout.DefinitionForView (editor.TextLabel);
				editor.GridLayout.SetWidth (0, titleDef.Column.Int32Value);
				editor.Style.EditorOffset = new UIOffset (10, 0);
			}

			if (property.Name == "Origin") {
				editor.Style.EditorOffset = new UIOffset (0, 0);
				editor.Style.SeparatorColor = null;
			}

			if (property.Name == "Name") {
				editor.Style.FeedbackLabelOffset = new UIOffset (10, 0);
				editor.FeedbackLabel.Font = UIFont.ItalicSystemFontOfSize (10);
			}

			if (property.Name == "Guests") {
				TKGridLayoutCellDefinition labelDef = editor.GridLayout.DefinitionForView (((TKDataFormStepperEditor)editor).ValueLabel);
				labelDef.ContentOffset = new UIOffset (-25, 0);
			}

			if (property.Name == "Section") {
				UIImage img = UIImage.FromBundle ("guest-name.png");
				editor.Style.TextLabelOffset = new UIOffset (img.Size.Width + 10, 0);
			}

			if (property.Name == "Table" || property.Name == "Section") {
				editor.TextLabel.TextColor = UIColor.White;
				editor.BackgroundColor = UIColor.Clear;
				((TKDataFormOptionsEditor)editor).SelectedOptionLabel.TextColor = UIColor.White;
				((TKDataFormOptionsEditor)editor).SelectedOptionLabel.TextAlignment = UITextAlignment.Right;
				editor.Style.EditorOffset = new UIOffset (-10, 0);
			}
		}

		public override void UpdateGroupView (TKDataForm dataForm, TKEntityPropertyGroupView groupView, uint groupIndex)
		{
			groupView.TitleView.TitleLabel.TextColor = UIColor.LightGray;
			groupView.TitleView.TitleLabel.Font = UIFont.SystemFontOfSize (13);
			groupView.TitleView.Style.Insets = new UIEdgeInsets (0, 10, 0, 0);
			if (groupIndex == 1) {
				groupView.EditorsContainer.BackgroundColor = UIColor.Clear;
			}
		}
			
	}
}

File: TelerikUIExamplesInSwift2.0/TelerikUIExamplesInSwift/DataFormCustomization.swift
//
//  DataFormCustomization.swift
//  TelerikUIExamplesInSwift
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//


class DataFormCustomization: TKDataFormViewController {

    let dataSource = TKDataFormEntityDataSource()
    let reservationForm = ReservationForm()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        dataSource.selectedObject = reservationForm
        self.dataForm.dataSource = dataSource
        self.dataForm.backgroundColor = UIColor(patternImage: UIImage(named: "wood-pattern")!)
        
        self.dataForm.registerEditor(CallEditor.self, forProperty: dataSource.entityModel.propertyWithName("phone"))
        self.dataForm.registerEditor(CallEditor.self, forProperty: dataSource.entityModel.propertyWithName("cancelReservation"))
        self.dataForm.registerEditor(TKDataFormOptionsEditor.self, forProperty: dataSource.entityModel.propertyWithName("section"))
        self.dataForm.registerEditor(TKDataFormOptionsEditor.self, forProperty: dataSource.entityModel.propertyWithName("table"))
        self.dataForm.registerEditor(TKDataFormSegmentedEditor.self, forProperty: dataSource.entityModel.propertyWithName("origin"))
        
        dataSource.entityModel.propertyWithName("name").image = UIImage(named: "guest-name")
        dataSource.entityModel.propertyWithName("phone").image = UIImage(named: "phone")
        dataSource.entityModel.propertyWithName("date").image = UIImage(named: "calendar")
        dataSource.entityModel.propertyWithName("time").image = UIImage(named: "time")
        dataSource.entityModel.propertyWithName("guests").image = UIImage(named: "guest-number")
        dataSource.entityModel.propertyWithName("table").image = UIImage(named: "table-number")
        
        for property in dataSource.entityModel.properties() as! [TKDataFormEntityProperty] {
            if property.name == "section" || property.name == "table" {
                property.groupKey = "TABLE DETAILS";
            }
            else if property.name == "origin" || property.name == "cancelReservation" {
                property.groupKey = "ORIGIN";
            }
            else {
                property.groupKey = "RESERVATION DETAILS";
            }
        }
    }
    func cancelReservation() {
        let alert = TKAlert()
        
        alert.style.cornerRadius = 3;
        alert.title = "Cancel Reservation";
        alert.message = "Reservation Canceled!";
        
        alert.addActionWithTitle("OK") { (TKAlert alert, TKAlertAction action) -> Bool in
            return true
        }
        
        alert.show(true)
    }
    
    //MARK:- TKDataFormDelegate
    
    override func dataForm(dataForm: TKDataForm!, validateProperty propery: TKDataFormEntityProperty!, editor: TKDataFormEditor!) -> Bool {
        if propery.name == "name" {
            let value = propery.value() as! NSString
            if value.length == 0 {
                return false
            }
        }
        
        return true
    }
    
    override func dataForm(dataForm: TKDataForm!, didValidateProperty propery: TKDataFormEntityProperty!, editor: TKDataFormEditor!) {
        if propery.name == "name" {
            if !propery.isValid {
                propery.feedbackMessage = "Please fill in the guest name"
            }
            else {
                propery.feedbackMessage = nil
            }
        }
    }
    
    override func dataForm(dataForm: TKDataForm!, updateEditor editor: TKDataFormEditor!, forProperty property: TKDataFormEntityProperty!) {
        editor.style.textLabelOffset = UIOffsetMake(25, 0)
        editor.style.separatorLeadingSpace = 40
        editor.style.accessoryArrowStroke = TKStroke(color: UIColor(red: 0.780, green: 0.2, blue: 0.223, alpha: 1.0))
        if property.name == "name" {
            if !property.isValid {
                editor.style.feedbackLabelOffset = UIOffsetMake(25, -5)
                editor.style.editorOffset = UIOffsetMake(25, -7)
            }
            else{
                editor.style.feedbackLabelOffset = UIOffsetMake(25, 0)
                editor.style.editorOffset = UIOffsetMake(25, 0)
            }
            
            editor.feedbackLabel.font = UIFont(name: "Verdana-Italic", size: 10)
            editor.style.textLabelDisplayMode = TKDataFormEditorTextLabelDisplayMode.Hidden
            (editor.editor() as! UITextField).placeholder = "Name"
        }
        
        if property.name == "time" {
            editor.style.textLabelDisplayMode = TKDataFormEditorTextLabelDisplayMode.Hidden;
            editor.style.textLabelOffset = UIOffsetMake(25, 0);
            (editor as! TKDataFormDatePickerEditor).dateFormatter.dateFormat = "h:mm a";
            (editor.editor() as! UIDatePicker).datePickerMode = UIDatePickerMode.Time;
        }
        
        if property.name == "date" {
            editor.style.textLabelDisplayMode = TKDataFormEditorTextLabelDisplayMode.Hidden
            editor.style.textLabelOffset = UIOffsetMake(25, 0)
        }
        
        if property.name == "guests" {
            let stepper = editor.editor() as! UIStepper
            stepper.minimumValue = 1
            stepper.tintColor = UIColor(red: 0.780, green: 0.2, blue: 0.223, alpha: 1.0)
            stepper.setIncrementImage(UIImage(named: "plus"), forState: UIControlState.Normal)
            stepper.setDecrementImage(UIImage(named: "minus"), forState: UIControlState.Normal)
        }
        
        if property.name == "section" {
            editor.textLabel.textColor = UIColor.whiteColor()
            editor.backgroundColor = UIColor.clearColor()
            (editor as! TKDataFormOptionsEditor).options = ["Section 1", "Section 2", "Section 3", "Section 4"]
            (editor as! TKDataFormOptionsEditor).selectedOptionLabel.textColor = UIColor.whiteColor()
        }
        
        if property.name == "table" {
            editor.textLabel.textColor = UIColor.whiteColor()
            editor.backgroundColor = UIColor.clearColor()
            (editor as! TKDataFormOptionsEditor).options = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15"]
            (editor as! TKDataFormOptionsEditor).selectedOptionLabel.textColor = UIColor.whiteColor()
        }
        if property.name == "origin" {
            editor.style.textLabelDisplayMode = TKDataFormEditorTextLabelDisplayMode.Hidden
            editor.style.editorOffset = UIOffsetMake(25, 0)
            (editor as! TKDataFormSegmentedEditor).segments = ["phone", "in-person", "online", "other"]
            let segmentedControl = editor.editor() as! UISegmentedControl
            segmentedControl.tintColor = UIColor(red:0.780, green:0.2, blue:0.223, alpha:1.0)
        }
        
        if property.name == "cancelReservation" {
            editor.style.textLabelDisplayMode = TKDataFormEditorTextLabelDisplayMode.Hidden
            editor.style.editorOffset = UIOffsetMake(25,0)
            (editor as! CallEditor).actionButton.setTitle(property.displayName, forState: UIControlState.Normal)
            (editor as! CallEditor).actionButton.addTarget(self, action: "cancelReservation", forControlEvents: UIControlEvents.TouchUpInside)
        }
    }
    
    override func dataForm(dataForm: TKDataForm!, viewForHeaderInSection section: Int) -> UIView! {
        let header = TKDataFormHeaderView()
        header.titleLabel.textColor = UIColor.grayColor()
        header.insets = UIEdgeInsetsMake(0, 40, 0, 0)
        header.separatorColor = TKSolidFill(color: UIColor.clearColor())
        if section == 0 {
            header.titleLabel.text = "RESERVATION DETAILS"
        }
        else if section == 1 {
            header.titleLabel.text = "TABLE DETAILS"
        }
        else{
            header.titleLabel.text = "ORIGIN"
        }
        
        return header
    }
    
    override func dataForm(dataForm: TKDataForm!, heightForHeaderInSection section: Int) -> CGFloat {
        return 30
    }
}


